daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala

```diff
@@ -1415,6 +1423,50 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
     }
   }
 
+  final def sortChildNodesByPosition(): Unit = {
+    // TODO: Once we upgrade to Scala 2.13 we can call the inplace sort method: DAFFODIL-2152
+    // childNodes = childNodes.sortInPlace(_.erd.position)
+    childNodes = childNodes.sortBy(_.erd.position)
+  }
+
+  /*
+   * When parsing unordered sequences it is possible to have non-contiguous arrays.
+   * When parsed, these arrays show up as separate DIArrays in the infoset. We need
+   * to combine these separate arrays.
+   * */
+  final def flattenAndValidateChildNodes(): Maybe[List[ElementRuntimeData]] = {
+    var ret: List[ElementRuntimeData] = List()
+    var groups = childNodes.groupBy(_.erd)
+    groups foreach {
+      case (erd, nodes) => {
+        // Check min/maxOccurs validity while iterating over childNodes
+        val min0 = erd.minOccurs
+        val max0 = erd.maxOccurs
+        val isUnbounded = max0 == -1
+
+        // Flatten multiple DIArrays into the first one
+        if (erd.isArray) {
+          nodes.reduceLeft((a, b) => a.asInstanceOf[DIArray].concat(b.asInstanceOf[DIArray]))
+          nodes.reduceToSize(1)
+        }
+
+        val occurrence = if (erd.isArray) nodes(0).contents.length else nodes.length
+        if (isUnbounded && occurrence < min0)
+          ret = ret :+ erd
+        else if (!isUnbounded && (occurrence < min0 || occurrence > max0))
+          ret = ret :+ erd
+      }
+    }
+
+    val (erds, nodes) = groups.unzip
+    childNodes.clear()
+    nodes.foreach( childNodes ++= _ )
+    if (ret.isEmpty)
+      Maybe.Nope
+    else
+      Maybe(ret)
+  }
+
   override def addChild(e: InfosetElement): Unit = {
     if (!e.isHidden && !hasVisibleChildren) hasVisibleChildren = true
     if (e.runtimeData.isArray) {
```

Comment:
So instead of actually flattening this, I updated the addChildToFastLookup to concat DIArrays, which resolves this issue and doesn't have any issues with backtracking.  I think there is already appropriate cleanup code in the restoreFrom function.

I have also added a test to verify this.

Reply:
I feel like there's an edge case where that wouldn't work. Imagine we parse an array, and then parse a scalar. Now say we try to parse another of the same array. WIth your change that will contact to the previous array. But then if that array element fails for some reason we need to be able to remove that array we concated. But I *think* the way we backtrack arrays is by looking the last element in the infoset and reducing the array size. But the array is not the last in the childNodes seq now, since the array we added is some previous array.